control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,101 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	GatewayClassFinalizer      = "gateway-exists-finalizer.networking.x-k8s.io"
+	GatewayClassControllerName = "hashicorp.com/consul-api-gateway-controller"
+)
+
+// GatewayClassReconciler reconciles a GatewayClass object.
+// The GatewayClass is responsible for defining the behavior of API gateways
+// which reference the given class.
+type GatewayClassReconciler struct {
+	ControllerName string
+	Log            logr.Logger
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/finalizers,verbs=update
+
+// Reconcile handles the reconciliation loop for GatewayClass objects.
+func (r *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("gatewayClass", req.NamespacedName)
+
+	gc := &gwv1beta1.GatewayClass{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, gc)
+	if err != nil {
+		log.Error(err, "unable to get GatewayClass")
```

Comment:
you'll want to do the `IsNotFound` check here or you're going to trigger infinite reconciliation if a gateway class has been fully GC'd.

Reply:
No worries! If it wasn't clear, I wouldn't be opposed to rewriting it to be more clear.